tools/chainsync-loadtest/src/concurrency/once.rs

```diff
@@ -0,0 +1,58 @@
+use std::future::Future;
+use std::sync::RwLock;
+
+fn is_send<T: Send>() {}
+fn is_sync<T: Sync>() {}
+
+#[allow(dead_code)]
+fn test<T: Clone + Send + Sync>() {
+    is_send::<Once<T>>();
+    is_sync::<Once<T>>();
+}
+
+// Once is a synchronization primitive, which stores a single value.
+// This value can be set at most once, and multiple consumers are
+// allowed to wait for that value.
+pub struct Once<T: Clone + Send + Sync> {
+    value: RwLock<Option<T>>,
+    notify: tokio::sync::Notify,
+}
+
+impl<T: Clone + Send + Sync> Once<T> {
+    pub fn new() -> Once<T> {
+        return Once { value: RwLock::new(None), notify: tokio::sync::Notify::new() };
+    }
+
+    // set() sets the value of Once to x and returns true.
+    // If the value was already set, it just returns false.
+    pub fn set(self: &Self, x: T) -> bool {
```

Comment:
Is it part of some style guide? I'd expect "cargo fmt" to do that for me if it was.

Reply:
I'm not sure if there's a style guide for this anywhere. `self`, `&self` and `&mut self` predate the ability to specify the self type at all, so for at least references everybody has been writing it the old way (both because its terser and because everybody else is doing it the old way.)

`rustfmt` fixing this up sounds like a good idea, but explicit self types are so obscure that this likely just never came up. Mind filing an issue against `rustfmt`, or should I?

Reply:
If you consider it worth it, please file a request. Like, IMHO converging towards self:&Self, rather than towards &self would make the syntax more uniform (since for wrapping types you have to state type explicitly anyway), so personally I don't know if it is a good idea in the first place. I was just curious whether this decision has already been "set in stone" in some sense.